### PR TITLE
feat: better errors for jquery and axios odata clients

### PR DESCRIPTION
**Pull request by Involture**
This is the first time I use the pull request feature on github so I hope I did everything correctly.

This PR brings the following changes:

1. Renamings
The goal is to make the origin and the type of the error clear
- rename axios-odata-client/RequestError -> axios-odata-client/AxiosODataClientError. 
- rename jquery-odata-client/RequestError -> jquery-odata-client/JQueryODataClientError.

2. Error types
- The underlying error (or any object) is linked to the cause property of errors
- If present, the response error message is linked to the root error object
- The error message is prefixed with a string indicating which kind of error it describes (from the server, from the client lib, other)

3. Testing
- Updates to integration testing to match the proposed changes.

Points to discuss :
- The status code could be lifted to the root error object too. It may be used often. However, some errors (bad client configuartion for example ?) may not have a status code, so it would be an optional field.
- Unit tests could be added for the client packages. That would mean simulating requests and responses from servers manually : storing the objects in test setup files. Inconvenient : Those samples requests and reponses would be hard to get right and up to spec. Advantage : right now with integration tests, if something breaks in the underlying client library or the test server, the tests would fail and wrongly suggest that something is wrong with the adapter. However I think that this will never happen. Testing in integration with the underlying clients and test servers is probably more reliable that manually writing requests and responses samples.

### CODE REVIEW

File: examples/bookshop/int-test/QueryFunctionality.test.ts [hunk 1]
```diff
@@ -39,23 +39,28 @@ describe("CAP V4 Integration Testing: Query Capabilities", () => {
   });
 
   test("get unknown book", async () => {
-    let failMsg = "Not Found";
-    await expect(() => testService.navToBooks().get(-1).query()).rejects.toThrow(failMsg);
+    let axiosClientMsgPrefix = "OData server error: ";
```
Comment (texttechne) [inline examples/bookshop/int-test/QueryFunctionality.test.ts:42]: both variables can be `const`. Was originally already wrong...

File: examples/odata-v2/int-test/ODataV2ServiceIntegration.test.ts [hunk 2]
```diff
@@ -49,23 +49,28 @@ describe("Integration Testing of generated stuff for Sample V2 OData Service", (
   });
 
   test("get unknown product", async () => {
-    let failMsg = "Resource not found for the segment 'Products'.";
-    await expect(() => testService.navToProducts().get(666).query()).rejects.toThrow(failMsg);
+    let axiosFailMsg = "Resource not found for the segment 'Products'.";
```
Comment (texttechne) [inline examples/odata-v2/int-test/ODataV2ServiceIntegration.test.ts:52]: `const` again please.

File: packages/axios-odata-client/src/AxiosODataClientError.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { AxiosError } from "axios";
+
+export class AxiosODataClientError extends Error {
+  name = "AxiosODataClientError";
+  isAxiosOdataClientError = true;
```
Comment (texttechne) [inline packages/axios-odata-client/src/AxiosODataClientError.ts:5]: do you think that this information "isAxisoOdataClientError" has any value? 
Should it be removed?

Or do you want to go for a type? Would be easier to determine which kind of error you have: either axios or Jquery..

File: packages/jquery-odata-client/src/JQueryODataClient.ts [hunk 4]
```diff
@@ -50,15 +50,10 @@ export class JQueryODataClient implements ODataClient<AjaxRequestConfig> {
...
-          reject(
-            Object.assign(new Error(), {
-              status: jqXHR.status || 0,
-              code: jqXHR.statusText,
-              data: jqXHR.responseJSON,
-              message,
-            } as RequestError)
-          );
+          if (message) {
+            reject(new JQueryODataClientError("OData server error: " + message, { cause: jqXHR }));
```
Comment (texttechne) [inline packages/jquery-odata-client/src/JQueryODataClient.ts:54]: either a  `return` or an `else` block => makes it clearer I think. At least I am thinking now, if this works or if `reject` gets called twice now....

**Comment by texttechne** [comment]
Great stuff! And good PR!

> The status code could be lifted to the root error object too. It may be used often. However, some errors (bad client configuartion for example ?) may not have a status code, so it would be an optional field.

Yes, I think that's a good point, please lift it. That's also something that should be common among all clients.

> Testing in integration with the underlying clients and test servers is probably more reliable that manually writing requests and responses samples

That was my thinking as well :-) So, I agree, let's stick with the integration tests.

File: packages/axios-odata-client/src/AxiosODataClient.ts [hunk 5]
```diff
@@ -104,24 +104,19 @@ export class AxiosODataClient implements ODataClient<AxiosRequestConfig> {
...
-      return Object.assign(new Error(), {
-        isRequestError: true,
-        stack: axiosError.stack,
-        message,
-        canceled: axios.isCancel(axiosError),
-        code: axiosError.code,
-        status,
-        data: axiosError.response ? axiosError.response.data : undefined,
-      });
+      if (message) {
```
Comment (texttechne) [inline packages/axios-odata-client/src/AxiosODataClient.ts:113]: I'm really not sure if this distinction makes sense. Well, at least based on the message....
If the server implements error messages wrong, then we would show an Axios Error now.

I would suggest to go with a default error message in this case, something like "Failed without message!" or something along this line.

**Comment by Involture** [comment]
Thank you for the review. I will integrate your remarks and update the PR (probably tomorrow morning)

**Comment by texttechne** [comment]
Hey @Involture,
do you have any troubles with this PR? Can I be of any help here?

I don't want to build up any pressure, just asking if I can help.

**Comment by Involture** [comment]
I forgot to run integration test, I think some of them won't pass.
One more commit incoming

**Comment by texttechne** [comment]
Thanks for taking this up again!

> Also I don't know if merging the main branch back in the PR when it is outdated is good practice or not. I figured it could save you the trouble of merging changes yourself

Actually, you have to do it this way. Perfect!

**Comment by Involture** [comment]
Everything should pass now.
What do you think of the changes I have made to the messages ? Is it more clear ?

**Comment by texttechne** [comment]
and merged: Thank you so much for this contribution!!!

**Pull request merged**